primitives/xcm/src/barriers.rs

```diff
@@ -77,3 +77,97 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionDes
 		}
 	}
 }
+
+/// Make sure that not withdrawable assets are handle properly according to the XCM design:
+/// - forbid any ReserveAssetDeposited instruction that contains not withdrawable asset(s) as
+/// parameter.
+/// - Morph some specific messages patterns that try to Widraw then Deposit not withdrawable
+/// asset(s).
+pub struct NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>(
+	core::marker::PhantomData<(IsWithdrawable, InnerBarrier)>,
+);
+
+impl<IsWithdrawable: crate::IsWithdrawable, InnerBarrier: ShouldExecute> ShouldExecute
+	for NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		InnerBarrier::should_execute(origin, instructions, max_weight, weight_credit)?;
+		let maybe_morphed_message = match &instructions.0[..] {
+			&[WithdrawAsset(ref assets), ClearOrigin, BuyExecution {
+				ref fees,
+				ref weight_limit,
+			}, DepositAsset {
+				assets: ref filter,
+				ref beneficiary,
+				..
+			}] => morph_xtokens_to_reserve_message::<Call, IsWithdrawable>(
+				assets,
+				fees,
+				weight_limit,
+				filter,
+				beneficiary,
+			),
+			_ => None,
+		};
+
+		if let Some(mut morphed_message) = maybe_morphed_message {
+			for i in 0..morphed_message.len() {
+				core::mem::swap(&mut instructions.0[i], &mut morphed_message[i]);
+			}
+		}
+
+		if instructions.0.iter().any(|instruction| {
+			matches!(
+				instruction, ReserveAssetDeposited(assets) if assets.inner().iter()
+				.any(|asset| IsWithdrawable::is_withdrawable(&asset))
+			)
+		}) {
+			Err(())
+		} else {
+			Ok(())
+		}
+	}
+}
+
+fn morph_xtokens_to_reserve_message<Call, IsWithdrawable: crate::IsWithdrawable>(
+	assets: &MultiAssets,
+	fees: &MultiAsset,
+	weight_limit: &WeightLimit,
+	deposit_filter: &MultiAssetFilter,
+	beneficiary: &MultiLocation,
+) -> Option<sp_std::vec::Vec<Instruction<Call>>> {
+	let mut not_withdrawable_assets = sp_std::vec::Vec::new();
+	let mut withdrawable_assets = sp_std::vec::Vec::new();
+	for asset in assets.inner() {
+		if IsWithdrawable::is_withdrawable(&asset) {
+			not_withdrawable_assets.push(asset.clone());
+		} else {
+			withdrawable_assets.push(asset.clone());
+		}
+	}
+	if not_withdrawable_assets.is_empty() {
+		None
+	} else {
+		Some(sp_std::vec![
+			WithdrawAsset(MultiAssets::from(withdrawable_assets)),
+			BuyExecution {
+				fees: fees.clone(),
+				weight_limit: weight_limit.clone(),
+			},
+			TransferAsset {
+				assets: MultiAssets::from(not_withdrawable_assets),
+				beneficiary: beneficiary.clone(),
+			},
+			DepositAsset {
```

Comment:
What would happen to weights that are calcuated for a set of instructrions, and then we change them essentially yielding another weight value?

Reply:
PS: I would still add the ClearOrigin after the TransferAsset

Reply:
I think the way you have it now works better. The innerBarrier goes before your message-modifier barrier, and therefore this means that the innerBarrier will pass with the proper set of instructions. We might still find some incompatibilities in the executor itself..

Reply:
> could we use setTopic from xcmV3 to mark which was the original origin?

I don't understand how it would help, we would still need to morph the incoming message, so what would be the advantage?